### PR TITLE
Fix bug #58400 - incorrect implementation of LocalPath for file URLs

### PR DESCRIPTION
**Pull request by kg**
https://bugzilla.xamarin.com/show_bug.cgi?id=58400

Uri.LocalPath should return a windows-style UNC path (```\\host\dir\filename```) for file URLs that have a hostname in them. Currently we have weird logic to return unix-style paths that intentionally (?) omits the hostname in most cases, which is extremely incorrect. This assumption was baked into some of our automated tests as well. There are a few corner cases involved that need to be handled to match Windows and .NET Core, for example that ```file://localhost/x/y``` is formally equivalent to ```file:///x/y``` in this case.

I considered that it might be more appropriate to try to return linux and OS X specific equivalents, because the MSDN documentation suggests that LocalPath is appropriate for the current OS. But Linux appears to have no UNC path equivalent, and OS X's ```smb://host/x/y``` is a weird and probably not appropriate thing to convert a file URI into. So I decided to just match Windows and .NET Core (on linux) behavior here.

### CODE REVIEW

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 1]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
         private string GetLocalPath(){
             EnsureParseRemaining();
 
+            // https://bugzilla.xamarin.com/show_bug.cgi?id=58400
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: move #if MONO here

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 2]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
+            bool isFileUrlWithHost = (
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Redundant parens

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 3]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
+                (m_Info.Offset.Host != m_Info.Offset.Path) &&
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Redundant parens

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 4]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
 
+            // https://bugzilla.xamarin.com/show_bug.cgi?id=58400
+            bool isFileUrlWithHost = (
+                (m_Info.Offset.Host != m_Info.Offset.Path) && 
+                // only file URLs
+                IsFile && 
+                // Paths that got reinterpreted as files would be treated
+                //  as if they have a host, like "/a/b" which maps to "file:///a/b/",
+                //  or "C:\a\b\" which maps to "file:///c:/a/b"
+                OriginalString.StartsWith("file://") &&
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Use StringComparer.Ordinal

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 5]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
+                // file://localhost/x/y needs to produce /x/y
+                !IsLoopback
+            );
+
+            // Manually check for "hosts" that are just forward slashes.
+            if (isFileUrlWithHost) {
+                isFileUrlWithHost = false;
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Use string.IndexOf look up instead

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 6]
```diff
@@ -1069,11 +1069,49 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
+            // We need to force generation of a UNC-style path instead of a unix one,
+            // despite the path not being UNC originally
+            bool treatAsUncPath = IsUncPath || isFileUrlWithHost;
+
 #if MONO
             //
             // I think this is wrong but it keeps LocalPath fully backward compatible
             //
-            if (IsUncOrDosPath && (IsWindowsFileSystem || !IsUncPath))
+            if (
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Remove unneeded parens and after that one line should be more readable

File: mcs/class/referencesource/System/net/System/URI.cs (hunk 5 above)
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Use string.IndexOf look up

File: mcs/class/referencesource/System/net/System/URI.cs (hunk 6 above)
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: I'd make it one line, this is hard to read

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 7]
```diff
@@ -1083,7 +1121,11 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
                 int start;
 
                 // Do we have a valid local path right in m_string?
-                if (NotAny(Flags.HostNotCanonical|Flags.PathNotCanonical|Flags.ShouldBeCompressed)) {
+                if (
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]: Should be under #if MONO

File: mcs/class/System/Test/System/UriTest3.cs [hunk 8]
```diff
@@ -696,6 +696,39 @@ public void DomainLabelLength ()
 			type = Uri.CheckHostName ("3.141592653589793238462643383279502884197169399375105820974944592._om");
 			Assert.AreEqual (UriHostNameType.Unknown, type, "DomainLabelLength#3");
 		}
+
+        // https://bugzilla.xamarin.com/show_bug.cgi?id=58400
+        [Test]
+        public static void Test_LocalPath_Bug58400()
+        {
+            var uriAndExpected = new global::System.Collections.Generic.Dictionary<string, string>
```
Comment (marek-safar) [inline mcs/class/System/Test/System/UriTest3.cs]: Could be just an array of named tuples

Comment (kg) [reply mcs/class/System/Test/System/UriTest3.cs]: Would be more code, wouldn't it? Because of the new Tuple<,>() per item

File: mcs/class/referencesource/System/net/System/URI.cs (hunk 5 above)
Comment (kg) [reply mcs/class/referencesource/System/net/System/URI.cs]: I don't think I can use ```IndexOf``` to do a ```!=``` like this

File: mcs/class/referencesource/System/net/System/URI.cs (hunk 6 above)
Comment (kg) [reply mcs/class/referencesource/System/net/System/URI.cs]: The precedence seems super unclear and easy to break without them. I got it wrong the first time. Can change to one line if you want

File: mcs/class/System/Test/System/UriTest3.cs (hunk 8 above)
Comment (marek-safar) [reply mcs/class/System/Test/System/UriTest3.cs]: C# supports e.g. `(value: string, data: string)` syntax for tuples

File: mcs/class/referencesource/System/net/System/URI.cs [hunk 9]
```diff
@@ -1083,7 +1122,15 @@ private static void SetEscapedDotSlashSettings(UriSectionInternal uriSection, st
...
+                    NotAny(Flags.HostNotCanonical|Flags.PathNotCanonical|Flags.ShouldBeCompressed) && 
+#if MONO
```
Comment (marek-safar) [inline mcs/class/referencesource/System/net/System/URI.cs]:
```
#if MONO
&& !isFileUrlWithHost
#endif
```

would work too
